src/image/ExposureFitsReader.cc

```diff
@@ -44,6 +44,17 @@ class ExposureFitsReader::MetadataReader {
 public:
     MetadataReader(std::shared_ptr<daf::base::PropertyList> primaryMetadata,
                    std::shared_ptr<daf::base::PropertyList> imageMetadata, lsst::geom::Point2I const& xy0) {
+        int version;
+        if (primaryMetadata->exists(ExposureInfo::getFitsSerializationVersionName())) {
+            version = primaryMetadata->getAsInt(ExposureInfo::getFitsSerializationVersionName());
+            primaryMetadata->remove(ExposureInfo::getFitsSerializationVersionName());
+        } else {
+            version = 0;  // unversioned files are implicitly version 0
+        }
+        if (version > ExposureInfo::getFitsSerializationVersion()) {
+            throw LSST_EXCEPT(pex::exceptions::TypeError, "Cannot read Exposure FITS version >= MAX_VERSION");
+        }
```

Comment:
I'm a bit torn about whether we should throw this exception here; doing so declares that there can never be a forward-compatible version increment, while it seems that being forward-compatible is something we could do if it happens to be easy.  The alternative would be to defer the exception until the version compatibility caused a problem, and perhaps even then set an unrecognized component to empty and warn (when possible), if doing so permits the rest of the exposure to be read in.  That's what we already do when (e.g.) the PSF component cannot be read because the package that defines it wasn't `setup`, and it's what usually happens now (without the warning) if you currently use old code to read a newer `Exposure` that has an unrecognized component.

That may not be possible when we're talking about replacing a component rather than adding one, but maybe we should just warn here until we run into a case where we have to throw?

Reply:
If I keep my promises, the next version of ExposureInfo will replace Calib with PhotoCalib and will thus make forward compatibility impossible from 0->1. If a particular forward-compatibility is possible, we could generalize this code; say, make it maxVersion+1 for that particular change. I feel like we are causing ourselves problems by always trying to "work but warn" and then having things fail much further down then line, when they should have failed early (i.e. DM-14310).

I wonder if we should bring in @leannep or @jdswinbank for management-level commentary?

Reply:
They're of course welcome to comment, but I'm happy to go with your preference on this; I certainly agree that the next anticipated version bump will not permit this kind of forward compatibility, and that's fine.